Extension/src/Debugger/configurationProvider.ts

```diff
@@ -505,17 +503,108 @@ export class CppConfigurationProvider implements vscode.DebugConfigurationProvid
+
+        let selection: MenuItem | undefined;
+
+        if (defaultConfig.length !== 0) {
+            selection = { label: defaultConfig[0].name, configuration: defaultConfig[0], description: defaultConfig[0].detail, detail: defaultConfig[0].existing };
+        } else {
+            let sortedItems: MenuItem[] = [];
+            // Find the recently used task and place it at the top of quickpick list.
+            const recentTask: MenuItem[] = items.filter(item => (item.configuration.preLaunchTask && item.configuration.preLaunchTask === DebugConfigurationProvider.recentBuildTaskLableStr));
+            if (recentTask.length !== 0) {
+                recentTask[0].detail = TaskConfigStatus.recentlyUsed;
+                sortedItems.push(recentTask[0]);
+            }
+            sortedItems = sortedItems.concat(items.filter(item => item.detail === TaskConfigStatus.configured));
+            sortedItems = sortedItems.concat(items.filter(item => item.detail === TaskConfigStatus.detected));
+            selection = await vscode.window.showQuickPick(sortedItems, {
+                placeHolder: (items.length === 0 ? localize("no.compiler.found", "No compiler found") : localize("select.debug.configuration", "Select a debug configuration"))
+            });
+        }
+
+        const debuggerEvent: string = DebuggerEvent.launchPlayButton;
+        if (!selection) {
+            Telemetry.logDebuggerEvent(debuggerEvent, { "debugType": debugModeOn ? "debug" : "run", "folderMode": folder ? "folder" : "singleMode", "cancelled": "true" });
+            return; // User canceled it.
+        }
+        if (!this.isClAvailable(selection.label)) {
+            return;
+        }
+        let resolvedConfig: vscode.DebugConfiguration | undefined | null;
+        if (selection.configuration && selection.configuration.type) {
+            resolvedConfig = await this.resolveDebugConfiguration(folder, selection.configuration);
+            if (resolvedConfig) {
+                resolvedConfig = await this.resolveDebugConfigurationWithSubstitutedVariables(folder, resolvedConfig);
+            }
+        }
+        if (resolvedConfig) {
+            await this.startDebugging(folder, resolvedConfig, debuggerEvent, debugModeOn);
+        }
     }
-}
 
-export class CppDbgConfigurationProvider extends CppConfigurationProvider {
-    public constructor(provider: IConfigurationAssetProvider) {
-        super(provider, DebuggerType.cppdbg);
+    private async startDebugging(folder: vscode.WorkspaceFolder | undefined, configuration: vscode.DebugConfiguration, debuggerEvent: string, debugModeOn: boolean = true): Promise<void> {
+
+        const debugType: string = debugModeOn ? "debug" : "run";
+        const folderMode: string = folder ? "folder" : "singleMode";
+        if (configuration.preLaunchTask) {
```

Comment:
I think it makes sense to separate this out from `startDebugging`.  It can either be its own method, or put it back in `buildAndDebug`